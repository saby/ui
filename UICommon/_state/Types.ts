export const SERIALIZED_KEY = '$serialized$';

export const TYPE = {
    INST: 'inst',
    FUNC: 'func',
    LINK: 'link',
    P_INF: '+inf',
    M_INF: '-inf',
    UNDEF: 'undef',
    NAN: 'NaN',
};

export const NEW_TYPE = {
    P_INF: '$+i',
    M_INF: '$-i',
    UNDEF: '$u',
    NAN: '$n',
};