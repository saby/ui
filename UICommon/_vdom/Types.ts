/**
 * Модуль, предоставляющий базовые типы опций и методы для работы с ними.
 */

import type { IVersionable } from 'Types/entity';

//# region Type defenitions

export interface IVersionableArray<T> extends Array<T> {
    getArrayVersion(): number;
    _arrayVersion: number;
}

export declare type TBuiltIn =
    | Date
    | Map<unknown, unknown>
    | Set<unknown>
    | WeakMap<object, unknown>
    | WeakSet<object>
    | object
    | symbol
    | number
    | string
    | boolean
    | null
    | undefined;

export declare type TExtended =
    | TBuiltIn
    | IVersionable
    | IVersionableArray<IOptions>
    | TContentOption;

export interface IOptions extends Record<string, TExtended> {
    [property: string]: TExtended | IOptions;
}

export declare type TControlOptionsExtended = IOptions | null | undefined;

export declare type TInternalsCollection = Map<number, TExtended | IOptions>;

export declare type TVersionsCollection = Map<string, number>;

//# endregion

//# region Template function interfaces

export interface TContentOptionFunction extends Function {
    internal: TInternalsCollection;
    isWasabyTemplate: boolean;
}

export interface TContentOptionObject extends Object {
    func: TContentOptionFunction;
    internal: TInternalsCollection;
    isWasabyTemplate: boolean;
}

export interface IDataObject extends Object {
    isDataArray: boolean;
    isWasabyTemplate: boolean;
}

export declare type TDataArray<T> = T[] & IDataObject;

export interface IFunctionDataArray<T> extends Function, IDataObject {
    array: T[];
}

/**
 * Тип шаблонной функции контентной опции tmpl.
 */
export interface TWS3ContentOption extends TContentOptionFunction {
    toJSON?(): Function;
}

/**
 * Тип шаблонной функции контентной опции wml (VDOM).
 */
export declare type TWS4VDOMContentOption = IFunctionDataArray<TContentOptionObject>;

/**
 * Тип шаблонной функции контентной опции wml (не VDOM).
 */
export declare type TWS4NVDOMContentOption = TDataArray<TContentOptionObject>;

/**
 * Тип шаблонной функции контентной опции wml.
 */
export declare type TWS4ContentOption = TWS4VDOMContentOption | TWS4NVDOMContentOption;

/**
 * Тип шаблонной функции контентной опции.
 */
export declare type TContentOption = TWS3ContentOption | TWS4ContentOption;

/**
 * Тип контентной опции, сгенерированной из React элемента children.
 */
export declare type TChildrenAsContentOptions = { isChildrenAsContent?: boolean };

//# endregion

//# region Helper functions

const UNREACHABLE_GETTER_PATH_FLAG = -100;

export function areBothNaN(a: unknown, b: unknown): boolean {
    return typeof a === 'number' && typeof b === 'number' && isNaN(a) && isNaN(b);
}

export function isContentOptionObject(value: unknown): value is TContentOptionObject {
    return !!(
        value &&
        typeof (value as TContentOptionObject).hasOwnProperty === 'function' &&
        (value as TContentOptionObject).hasOwnProperty('func') &&
        typeof (value as TContentOptionObject).func === 'function' &&
        (value as TContentOptionObject).hasOwnProperty('internal') &&
        (value as TContentOptionObject).hasOwnProperty('isWasabyTemplate')
    );
}

export function isWS3ContentOption(value: unknown): value is TWS3ContentOption {
    return (
        typeof value === 'function' &&
        typeof (value as TWS3ContentOption).hasOwnProperty === 'function' &&
        (value as TWS3ContentOption).hasOwnProperty('func') &&
        (value as TWS3ContentOption).hasOwnProperty('internal') &&
        (value as TWS3ContentOption).hasOwnProperty('isWasabyTemplate')
    );
}

export function getWS3ContentOptionInternals(value: TWS3ContentOption): TInternalsCollection {
    return value.internal;
}

export function isWS4NVDOMContentOption(value: unknown): value is TWS4NVDOMContentOption {
    return (
        Array.isArray(value) &&
        (value as TWS4NVDOMContentOption).isDataArray &&
        typeof (value as TWS4NVDOMContentOption).hasOwnProperty === 'function' &&
        (value as TWS4NVDOMContentOption).hasOwnProperty('isWasabyTemplate') &&
        (value as TWS4NVDOMContentOption).every(isContentOptionObject)
    );
}

export function isWS4VDOMContentOption(value: unknown): value is TWS4VDOMContentOption {
    return (
        !!value &&
        typeof value === 'function' &&
        (value as TWS4VDOMContentOption).isDataArray &&
        typeof (value as TWS4VDOMContentOption).hasOwnProperty === 'function' &&
        (value as TWS4VDOMContentOption).hasOwnProperty('isWasabyTemplate') &&
        Array.isArray((value as TWS4VDOMContentOption).array) &&
        (value as TWS4VDOMContentOption).array.every(isContentOptionObject)
    );
}

export function isWS4ContentOption(
    value: unknown
): value is TWS4NVDOMContentOption | TWS4VDOMContentOption {
    return isWS4NVDOMContentOption(value) || isWS4VDOMContentOption(value);
}

export function getWS4ContentOptionInternals(value: TWS4ContentOption): TInternalsCollection {
    if (isWS4NVDOMContentOption(value)) {
        // Контракт: контентная опция имеет строго 1 элемент
        return value[0].internal;
    }

    // Контракт: контентная опция имеет строго 1 элемент
    return value.array[0].internal;
}

export function isContentOption(value: unknown): value is TContentOption {
    return isWS4ContentOption(value) || isWS3ContentOption(value);
}

export function isVersionable(value: unknown): value is IVersionable {
    return !!(
        value &&
        typeof (value as IVersionable) === 'object' &&
        typeof (value as IVersionable).getVersion === 'function'
    );
}

export function isVersionableArray<T>(value: unknown): value is IVersionableArray<T> {
    return !!(
        value &&
        Array.isArray(value as IVersionableArray<T>) &&
        typeof (value as IVersionableArray<T>).getArrayVersion === 'function'
    );
}

export function isChildrenAsContent(value: unknown): boolean {
    return !!(value as TChildrenAsContentOptions).isChildrenAsContent;
}

export function shouldIgnoreChanging(value: IOptions | TExtended): boolean {
    return !!(value && (value as Record<string, boolean>)._ignoreChanging);
}

export function shouldCheckDeep(value: IOptions | TExtended): boolean {
    return !!(value && (value as Record<string, boolean>)._isDeepChecking);
}

export function shouldCheckVersions(value: IOptions | TExtended): boolean {
    return !!(value && (value as Record<string, boolean>)._preferVersionAPI);
}

export function shouldLazilyExit(value: TInternalsCollection): boolean {
    return value.has(UNREACHABLE_GETTER_PATH_FLAG);
}

export function installBlockOptionProperties(options: IOptions, blockOptionNames: string[]): void {
    for (const optionName of blockOptionNames) {
        const value = options[optionName];

        if (value && typeof options[optionName] === 'object') {
            Object.defineProperty(options[optionName], '_$blockOption', {
                value: true,
                enumerable: false,
            });
        }
    }
}

export function isBlockOption(value: TExtended | IOptions): boolean {
    return !!(value && (value as Record<string, boolean>)._$blockOption);
}

export function isPossiblyScopeObjectOfInternal(value: unknown): value is TControlOptionsExtended {
    return !!(
        value &&
        typeof value === 'object' &&
        // у scope есть _$internal
        (value as Record<string, boolean>)._$internal &&
        // We don't need to check Date object internal properties
        !(value instanceof Date) &&
        !isVersionable(value)
    );
}

//# endregion
