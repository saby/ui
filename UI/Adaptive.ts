/**
 * Библиотека для настройки адаптивности сайта к различным аспектам адаптивности
 * @see https://n.sbis.ru/wasaby/knowledge#toc_2eb25e0a-c867-45a8-a13d-b8c1ff12cdbc
 * @library UICore/Adaptive
 * @public публичная библиотека
 */

export {
    DEFAULT_BREAKPOINTS,
    AdaptiveContainer,
    AdaptiveInitializerConfig,
    AdaptiveModeType,
    useAdaptiveMode,
    withAdaptiveMode,
    getAdaptiveModeForLoaders,
    unsafe_getRootAdaptiveMode,
    ScrollOnBodyStore,
    BodyScroll,
    moveBodyScroll,
    unMoveBodyScroll,
    ScrollOnBodyContext,
    ScrollOnBodyContextProvider,
} from 'UICore/Adaptive';
