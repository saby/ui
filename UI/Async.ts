/**
 * Библиотека методов для асинхронной загрузки контролов.
 * @library UI/Async
 * @public
 * @author Мустафин Л.И.
 */

export { default as Async } from './_async/Async';
export { lazy, importer, ILazyProps } from './_async/lazy';
