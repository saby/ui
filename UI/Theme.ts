export { IActiveTheme } from 'UI/theme/controller';
export { default as ThemeBackground } from 'UI/_theme/ThemeBackground';
export { default as BackgroundViewer, isLight } from 'UI/_theme/BackgroundViewer';
export { IBackground } from 'UI/_theme/background/IBackground';
export { ThemeContext } from 'UI/_theme/context';
export { default as ThemeConsumer } from 'UI/_theme/Consumer';
export { default as ThemeDesigner } from 'UI/_theme/Designer';
export { default as getThemeStorage } from 'UI/_theme/ThemeStorage';
export { default as ThemeWrapper, ERROR_FALLBACK_SELECTOR } from 'UI/_theme/Wrapper';
export type { IThemeWrapperOptions, IErrorActiveTheme } from 'UI/_theme/Wrapper';
export { useActiveTheme } from 'UI/_theme/useActiveTheme';
