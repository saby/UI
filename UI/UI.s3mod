<?xml version="1.0" encoding="UTF-8" ?>
<ui_module id="f27abe12-19df-4c9f-b886-634bd20a0011" name="UI" package="Платформа" responsible="Шипин А.А." required="1" version="1.00">

  <description>Библиотека, позволяющая собрать "каркас" веб-приложения; содержит набор правил, техник и решений, допустимых при разработке прикладных решений.</description>

  <depends>
    <module id="0df0fce3-53ba-47f8-987d-edf1fd078295" name="WS.Core"/>
    <module id="10efeac7-0f02-4c73-9474-0d1d0db077b6" name="View"/>
    <module id="636177e5-4d75-44fa-9173-977dc90ff111" name="Env"/>
    <module id="0f5d26db-dd61-4957-9375-8861ad66f219" name="Vdom"/>
    <module id="63e1a7e5-4d75-42fa-91a3-977dc90ff1e1" name="SbisEnv"/>
    <module id="323fd54f-1fab-42a6-8aea-6726d6392d53" name="Application"/>
    <module id="832cda0a-e19c-4ffc-a1d1-47edbb6b7501" name="I18n"/>
  </depends>

  <load_after/>

  <parameters/>

  <resources/>

  <preload/>

</ui_module>
