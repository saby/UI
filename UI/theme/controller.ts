/// <amd-module name='UI/theme/controller' />
/**
 * Библиотека контроллера тем
 * @remark
 * Контроллер управляет загрузкой/удаленим темизированных стилей на wasaby-странице
 * {@link Темы https://wi.sbis.ru/doc/platform/developmentapl/interface-development/themes/}
 * @library UI/theme/controller
 * @includes getThemeController UI/theme/_controller/Controller#getInstance
 * @includes THEME_TYPE UI/theme/_controller/CssLink#THEME_TYPE
 * @public
 * @author Ибрагимов А.А.
 */

export { getInstance as getThemeController } from 'UI/theme/_controller/Controller';
export { THEME_TYPE } from 'UI/theme/_controller/css/Base';
