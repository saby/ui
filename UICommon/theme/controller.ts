/**
 * Библиотека контроллера тем
 * @remark
 * Контроллер управляет загрузкой/удалением стилей всех контроллов на wasaby-странице на клиенте и СП
 * Подробнее о работе с темами оформления читайте {@link https://wi.sbis.ru/doc/platform/developmentapl/interface-development/themes/ здесь}.
 * @library UICommon/theme/controller
 * @includes getThemeController UICommon/theme/_controller/Controller#getInstance
 * @includes THEME_TYPE UICommon/theme/_controller/css/const#THEME_TYPE
 * @includes EMPTY_THEME UICommon/theme/_controller/css/const#EMPTY_THEME
 * @includes DEFAULT_THEME UICommon/theme/_controller/css/const#DEFAULT_THEME
 * @public
 */
import { Controller } from './_controller/Controller';
import LinkResolver from './_controller/LinkResolver';
import Loader from './_controller/Loader';

export const getThemeController = Controller.getInstance;
export { TScopeType, TSiteType, TClassList, IActiveTheme } from './_controller/Controller';
export { THEME_TYPE, EMPTY_THEME, DEFAULT_THEME } from './_controller/css/const';
export { LinkResolver, Controller, Loader };
