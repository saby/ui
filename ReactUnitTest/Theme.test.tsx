/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { Model } from 'Types/entity';
import { ThemeWrapper } from 'UI/Theme';
import { IActiveTheme, getThemeController, THEME_APPLICATIONS } from 'UI/theme/controller';
describe('Темизация', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div');
    });

    afterEach(() => {
        jest.restoreAllMocks();
        unmountComponentAtNode(container);
        container.remove();
    });

    describe('ThemeWrapper', () => {
        it('При первом рендере применяется активная тема из пропсов', () => {
            const applyStylesSpy = jest.spyOn(getThemeController(), 'applyStyles');
            const getActiveThemeSpy = jest.spyOn(getThemeController(), 'getActiveTheme');
            const className = 'activeThemeClass';
            const activeTheme: IActiveTheme = {
                selector: 'activeThemeSelector',
                version: 1,
                classList: [className],
            };

            act(() => {
                render(
                    <ThemeWrapper activeTheme={activeTheme}>
                        <div>test</div>
                    </ThemeWrapper>,
                    container
                );
            });

            expect(document.body.classList).toContain(className);
            expect(applyStylesSpy).toBeCalledWith(activeTheme);
            expect(getActiveThemeSpy).not.toBeCalled();
        });
        it('Если ThemeWrapper не верхний, список классов он навешивает на children', () => {
            const topActiveTheme: IActiveTheme = {
                classList: ['activeThemeClass'],
            };

            const bottomActiveTheme: IActiveTheme = {
                classList: ['bottomActiveThemeClass'],
                selector: 'default',
            };

            act(() => {
                render(
                    <ThemeWrapper activeTheme={topActiveTheme}>
                        <div className="themeWrapperChildren">
                            <ThemeWrapper activeTheme={bottomActiveTheme}>
                                <div className="bottomThemeWrapperChildren">test</div>
                            </ThemeWrapper>
                        </div>
                    </ThemeWrapper>,
                    container
                );
            });

            expect(container).toMatchSnapshot();
        });

        it('Применяется статическая тема, указанная в активной', () => {
            const staticTheme = 'default__test_min';
            const activeTheme: IActiveTheme = {
                selector: 'default__test',
                version: 1,
                classList: ['t-default__test'],
                themeApply: THEME_APPLICATIONS.palette,
                properties: Model.fromObject<IActiveTheme>({
                    staticTheme,
                }),
            };
            const arrActiveTheme = [activeTheme];

            act(() => {
                render(
                    <ThemeWrapper activeTheme={activeTheme}>
                        <div className="themeWrapperChildren">
                            <ThemeWrapper activeTheme={arrActiveTheme}>
                                <div className="bottomThemeWrapperChildren">test</div>
                            </ThemeWrapper>
                        </div>
                    </ThemeWrapper>,
                    container
                );
            });

            expect(document.body.className).toContain(staticTheme);
            expect(container).toMatchSnapshot();
        });
    });
});
