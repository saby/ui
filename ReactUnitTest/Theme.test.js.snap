// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Темизация ThemeWrapper Если ThemeWrapper не верхний, список классов он навешивает на children 1`] = `
<div>
  <div
    class="themeWrapperChildren controls_themes__wrapper"
  >
    <div
      class="bottomThemeWrapperChildren controls_theme-default bottomActiveThemeClass controls_themes__wrapper"
    >
      test
    </div>
  </div>
</div>
`;

exports[`Темизация ThemeWrapper Применяется статическая тема, указанная в активной 1`] = `
<div>
  <div
    class="themeWrapperChildren controls_themes__wrapper"
  >
    <div
      class="bottomThemeWrapperChildren controls_theme-default__test_min t-default__test controls_themes__wrapper"
    >
      test
    </div>
  </div>
</div>
`;
